services/typeddata/hash_test.go

```diff
@@ -0,0 +1,261 @@
+package typeddata
+
+import (
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTypeString(t *testing.T) {
+	type testCase struct {
+		description string
+		typeString  string
+		types       Types
+		target      string
+	}
+	for _, tc := range []testCase{
+		{
+			"WithoutDeps",
+			"Person(string name,address wallet)",
+			Types{"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}}},
+			"Person",
+		},
+		{
+			"SingleDep",
+			"Mail(Person from,Person to)Person(string name,address wallet)",
+			Types{
+				"Person": []Field{{Name: "name", Type: "string"}, {Name: "wallet", Type: "address"}},
+				"Mail":   []Field{{Name: "from", Type: "Person"}, {Name: "to", Type: "Person"}},
+			},
+			"Mail",
+		},
+		{
+			"DepsOrdered",
+			"Z(A a,B b)A(string name)B(string name)",
+			Types{
+				"A": []Field{{Name: "name", Type: "string"}},
+				"B": []Field{{Name: "name", Type: "string"}},
+				"Z": []Field{{Name: "a", Type: "A"}, {Name: "b", Type: "B"}},
+			},
+			"Z",
+		},
+		{
+			"RecursiveDepsIgnored",
```

Comment:
You mean `RecursiveDepsWorks`? It seems like they are not ignored and the output is generated as expected.

Reply:
what i meant is that A has a Z as a dependency, but it was already included and therefore ignored